src/kernel-install/kernel-install

```diff
@@ -89,44 +94,75 @@ elif [ -r "/usr/lib/kernel/install.conf" ]; then
     . /usr/lib/kernel/install.conf
 fi
 
-# Prefer to use an existing machine ID from /etc/machine-info or /etc/machine-id. If we're using the machine
-# ID /etc/machine-id, try to persist it in /etc/machine-info. If no machine ID is found, try to generate
-# a new machine ID in /etc/machine-info. If that fails, use "Default".
-[ -z "$MACHINE_ID" ] && [ -r /etc/machine-info ]              && . /etc/machine-info && MACHINE_ID="$KERNEL_INSTALL_MACHINE_ID"
-[ -z "$MACHINE_ID" ] && [ -r /etc/machine-id ]                && read -r MACHINE_ID </etc/machine-id
-[ -n "$MACHINE_ID" ] && [ -z "$KERNEL_INSTALL_MACHINE_ID" ]   && echo "KERNEL_INSTALL_MACHINE_ID=$MACHINE_ID" >>/etc/machine-info
-[ -z "$MACHINE_ID" ] && NEW_MACHINE_ID="$(systemd-id128 new)" && echo "KERNEL_INSTALL_MACHINE_ID=$NEW_MACHINE_ID" >>/etc/machine-info
-[ -z "$MACHINE_ID" ] && [ -r /etc/machine-info ]              && . /etc/machine-info && MACHINE_ID="$KERNEL_INSTALL_MACHINE_ID"
-[ -z "$MACHINE_ID" ] && MACHINE_ID="Default"
-
-[ -z "$BOOT_ROOT" ] && for suff in "$MACHINE_ID" "loader/entries"; do
-    for pref in "/efi" "/boot" "/boot/efi" ; do
+# If /etc/machine-id is initialized we'll use it, otherwise we'll use a freshly
+# generated one. If the user configured an explicit machine ID to use in
+# /etc/machine-info to use for our purpose, we'll use that instead (for
+# compatibility).
+[ -z "$MACHINE_ID" ] && [ -r /etc/machine-info ] && . /etc/machine-info && MACHINE_ID="$KERNEL_INSTALL_MACHINE_ID"
+[ -z "$MACHINE_ID" ] && [ -r /etc/machine-id ]   && read -r MACHINE_ID </etc/machine-id
+[ -z "$MACHINE_ID" ] && MACHINE_ID="$(systemd-id128 new)"
```

Comment:
Shouldn't we drop this if we're not going to persist it? We'll end up creating a directory in /efi, /boot with no way to trace it back to a specific installation.

Reply:
But we *do* persist it! But not in /etc but in the generated boot entry snippets, i.e. via the systemd.machine_id= kernel cmdline switch! That way, if you do the factory reset all of /etc/ can be flushed out, but you will still get the same machine ID once the machine is back up, and thus you will then also be able to recognize the generated boot entry snippets again.

That's the magic of it: machine is persisted without being persisted ;-)

Reply:
What I'm worried about is that this ends up coupling the machine ID to the installation directory in the ESP quite heavily. Unless mkosi starts writing /etc/kernel/entry-token, images will end up with a systemd.machine_id snippet in the kernel cmdline, which will cause all instances of the image to have the same machine ID suddenly. The way I understand it, the most likely scenario where a machine ID won't be available is when building an image for distribution and that's exactly the scenario where we don't want to persist the machine ID to be persisted but instead want to have it generated on boot. So I figure most scenarios where this logic ends up being executed will be scenarios where we don't actually want it to execute (and the reason it'll be executed will be because someone unfamiliar with the space won't have written /etc/kernel/entry-token). 

The ideal behavior to me would be to have stuff get written under the IMAGE_ID/ID dir if IMAGE_ID/ID is configured, or otherwise "Default" if no machine ID is available when building a golden image.

Reply:
> The ideal behavior to me would be to have stuff get written under the IMAGE_ID/ID dir if IMAGE_ID/ID is configured, or otherwise "Default" if no machine ID is available when building a golden image.

Well, yes, gold image builders should do that. But we cannot auto-detect this case, and so far we always use the real machine-id. This change cleans things up, and allows you to clearly pick the right behaviour.

> So to summarize, I'm worried that if we ship this like is, we'll end up breaking compat for image builders since suddenly all instances of an image will end up with a fixed machine ID if you run kernel-install with no machine ID available and nothing in /etc/kernel/entry-token.

But you already end up with exactly that because we use the machine ID in the path to the snippets! So nothing is lost here.

Reply:
I might be wrong here, but what I'm thinking of is the following scenario where we do type 1 images in mkosi (without the 250 stuff):

- We configure a machine ID during the image build process because dracut doesn't work without one
- kernel-install runs as a post install script during the image build process and uses the machine ID we generated to install stuff to /efi
- We remove /etc/machine-id at the end of the build process so the generated image doesn't ship with a machine ID configured. When an instance of the image boots, the machine ID gets generated on boot.
- If kernel-install is run again after the first boot, it creates a new directory in /efi named after the final machine ID and installs stuff under that. We end up with 2 directories in /efi containing stuff from kernel-install.

What will happen with this patch included and without changes to mkosi:

- We configure a machine ID during the image build process because dracut doesn't work without one
- kernel-install runs as a post install script during the image build process and uses the machine ID we generated to install stuff to /efi and also includes systemd.machine_id in the kernel cmdline with the machine ID we generated
- We remove /etc/machine-id at the end of the build process so the generated image doesn't ship with a machine ID configured. When an instance of the image boots, the machine ID configured in systemd.machine_id gets used as the machine ID. This will be the same for every instance of the image that's booted.
- If kernel-install is run again after the first boot, it creates a new directory in /efi named after the final machine ID and installs stuff under that. Since the final machine ID is the same as the initial one we generated in mkosi, we end up with a single directory under /efi containing machine ID stuff.

What will happen with this patch included and with changing mkosi to have no machine ID during the build process:

- kernel-install runs as a post install script during the image build process and generates a random machine ID to install stuff to /efi and also includes systemd.machine_id in the kernel cmdline with the machine ID we generated
- We remove /etc/machine-id at the end of the build process so the generated image doesn't ship with a machine ID configured. When an instance of the image boots, the machine ID configured in systemd.machine_id gets used as the machine ID. This will be the same for every instance of the image that's booted.
- If kernel-install is run again after the first boot, it creates a new directory in /efi named after the final machine ID and installs stuff under that. Since the final machine ID is the same as the initial one we generated in mkosi, we end up with a single directory under /efi containing machine ID stuff.

So what I'm worried about is that unless image builders very explicitly configure the directory to use by either running bootctl with the right option first or write to /etc/kernel/entry-token first, they'll end up with a very wrong configuration (same machine ID for every instance of booted images) that isn't immediately obvious and probably won't break things very noticeably in the beginning, but can be a cause for insidious bugs later on. And this stuff isn't immediately obvious to anyone working on an image builder unless they end up reading this discussion or we very clearly document this somewhere where people will end up reading it.

Previously, yeah we'd ship stuff in /efi under the wrong machine ID directory, but it wasn't noticeable because kernel-install would just create a new directory and put new stuff under that. And since we removed the machine-id from the filesystem, a new unique one would be generated for each instance on boot. With this patch, we'll actually end up shipping the same machine ID for every instance of the image.

Reply:
Quite frankly, given this was broken in any case (i.e. the orphaned entry dir we so far generate just sucks! and the fact we can't match up what the boot loader reports with what we look for on next book is a bad bug), I am not sure it would really be a regression. We just move broken behaviour from one place to another, which I think is not a total desaster, given the ultimate outcome is an improvement.

Note that for the specific case of mkosi the issue shouldn't matter, since there we generate type 2 entries anyway.

That said, if this is important to you I figure we could tweak this further:

1. a simple hack would be: don't write the machine ID into the generated entry file if we run in a container. that should work, but is slightly skewed I guess. Thunking behind it: we are apparently prepping an image in a container environment we intend to use in a baremetal or VM environment, otherwise we wouldn't bother with kernel-install. But if that's the case we can probably assume it's going to be a golden image situation.

2. another option is to only persist the machine ID in the kernel cmdline if `test -f /etc/kernel/entry-token` also holds, i.e. if the machine ID was *explicitly* configured to be used. This would thus mean the behaviour must be explicitly requested (i.e. the file must be created, and given the file is a new concept people won't get opted into this automatically just like that).

Reply:
I think if we add a clear note in the relevant documentation warning image builders that they need to take explicit action here to avoid the machine ID getting coded into the spec files should be sufficient. Users might still end up running into this but they should be able to figure things out if the problem is documented in the relevant documentation.

Reply:
I'll add a NEWS file for that.